editor/src/clj/editor/editor_extensions/vm.clj

```diff
@@ -229,3 +229,68 @@
   LuaTable (->clj [table vm]
              ;; only hold the lock for the mutable lua table access
              (with-lock vm (lua-table->clj-map-or-vector table vm))))
+
+(defn lua-table-reducer [vm ^LuaTable lua-table]
+  (reify IReduceInit
+    (reduce [_ f init]
+      (with-lock vm
+        (loop [prev-k LuaValue/NIL
+               acc init]
+          (let [varargs (.next lua-table prev-k)
+                lua-k (.arg1 varargs)]
+            (if (.isnil lua-k)
+              acc
+              (let [acc (f acc varargs)]
+                (if (reduced? acc)
+                  @acc
+                  (recur lua-k acc))))))))))
+
+(defn lua-value->string [vm lua-value]
+  (condp instance? lua-value
+    LuaString (pr-str (str lua-value))
+    LuaTable (let [r (lua-table-reducer vm lua-value)
+                   ;; max index is either:
+                   ;; - highest pos int iff all keys are pos ints
+                   ;; - zero
+                   max-index (long
+                               (transduce
+                                 (comp
+                                   (map (fn [^Varargs varargs]
+                                          (let [k (.arg1 varargs)]
+                                            (if (.isinttype k)
+                                              (.tolong k)
+                                              0))))
+                                   (halt-when (complement pos-int?)))
+                                 max
+                                 0
+                                 r))
+                   sb (StringBuilder. "{")
+                   sb-rf (completing #(.append ^StringBuilder %1 %2))]
+               (if (pos? max-index)
+                 (do (transduce
+                       (comp
+                         (map (fn [i]
+                                (lua-value->string vm (.rawget ^LuaTable lua-value (unchecked-inc-int (int i))))))
+                         (interpose ", "))
+                       sb-rf
+                       sb
+                       (range max-index)))
+                 (do (transduce
```

Comment:
It looks like we could unwrap both these expressions from `do`?

Reply:
Yes! I used to have `(.append sb "}")` on both, but then moved it after the `if`.